src/sentry/tagstore/snuba/backend.py

```diff
@@ -157,21 +162,50 @@ def __get_tag_keys_for_projects(
         limit=1000,
         keys=None,
         include_values_seen=True,
+        use_cache=True,
         **kwargs
     ):
-        filters = {"project_id": projects}
+        filters = {"project_id": sorted(projects)}
         if environments:
-            filters["environment"] = environments
+            filters["environment"] = sorted(environments)
         if group_id is not None:
             filters["group_id"] = [group_id]
         if keys is not None:
-            filters["tags_key"] = keys
+            filters["tags_key"] = sorted(keys)
         aggregations = [["count()", "", "count"]]
 
         if include_values_seen:
             aggregations.append(["uniq", "tags_value", "values_seen"])
         conditions = []
 
+        should_cache = (
+            random.random() <= options.get("snuba.tagstore.cache-tagkeys-rate")
+            and use_cache
+            and group_id is None
+        )
+
+        if should_cache:
+            filtering_strings = [
+                u"{}={}".format(key, value) for key, value in six.iteritems(filters)
+            ]
+            cache_key = u"testing.tagstore.__get_tag_keys:{}".format(
+                md5_text(*filtering_strings).hexdigest()
+            )
+            # Round times by 5 minutes since we cache for 5 minutes
+            if start:
+                start = start.replace(minute=start.minute // 5 * 5, second=0, microsecond=0)
+                cache_key += ":" + start.isoformat()
+            if end:
+                end = end.replace(minute=end.minute // 5 * 5, second=0, microsecond=0)
+                cache_key += ":" + end.isoformat()
+
+            cache_result = cache.get(cache_key, False)
+            if cache_result:
+                metrics.incr("testing.tagstore.cache_tag_key.hit")
+            else:
+                metrics.incr("testing.tagstore.cache_tag_key.miss")
+                cache.set(cache_key, True, 300)
```

Comment:
Also, since you are rolling this out slowly, you can directly store the result in cache and actually use the cache instead of storing the boolean. We just need to start low.

Reply:
I was going to go with 1% but should I go even lower?

Reply:
1 should be fine